### DISCUSSION TITLE
How to generate datasets for inductive learning?

### DISCUSSION THREAD

**Discussion by scottshufe** [pyg-team/pytorch_geometric#3415] [Q&A] [answered]
Hi, everyone. My question is how to generate datasets for inductive learning?
Using **_Cora_** dataset as an example, I want to use half of its nodes and their links for training, the rest for testing. I am not sure if this approach is correct:
1. randomly sample half of nodes on the graph
2. use the sampled nodes to generate subgraph(1) with torch_geometric.utils.subgraph function, the rest nodes for subgraph(2)
3. use the generated subgraph(1) as the training dataset and subgraph(2) for testing

Does anyone have better ideas?

**Comment by wsad1** [comment]
For datasets like cora I have mostly only seen examples of transductive learning (nodes in same graph are split into train/test. although this is not completely transductive as there is nothing stopping you from predicting on an unseen node) as done [here](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/cora.py). 

The approach you are suggesting sounds interesting, but because you are generating subgraphs, the neighborhood of a node might change, which might effect your prediction.

**Reply by scottshufe** [reply]
Got it. To make it simple, I just want to split a graph into two parts.
Thanks for your suggestion. :)

**Comment by rusty1s** [comment, marked as answer]
```python
train_mask = torch.rand(data.num_nodes) < 0.5
test_mask = ~train_mask

train_data = copy.copy(data)
train_data.edge_index, _ = subgraph(train_mask, data.edge_index, relabel_nodes=True)
train_data.x = data.x[train_mask]

test_data = copy.copy(data)
test_data.edge_index, _ = subgraph(test_mask, data.edge_index, relabel_nodes=True)
test_data.x = data.x[test_mask]
```

**Reply by scottshufe** [reply]
Thank you Matthias! I think this is what I need.
